src/openloco/window.cpp

```diff
@@ -315,6 +315,149 @@ namespace openloco::ui
         return scrollIndex;
     }
 
+    // 0x00459E54
+    // TODO: needs expansion in terms of (output) parameters.
+    static void get_map_coordinates_from_pos(int32_t screenX, int32_t screenY, int32_t flags, int16_t* x, int16_t* y)
+    {
+        registers regs;
+        regs.ax = screenX;
+        regs.cx = screenY;
+        regs.edx = flags;
+        call(0x00459E54, regs);
+
+        *x = regs.ax;
+        *y = regs.cx;
+    }
+
+    void window::viewport_get_map_coords_by_cursor(int16_t* map_x, int16_t* map_y, int16_t* offset_x, int16_t* offset_y)
+    {
+        // Get mouse position to offset against.
+        int32_t mouse_x, mouse_y;
+        ui::get_cursor_pos(mouse_x, mouse_y);
+
+        // Compute map coordinate by mouse position.
+        get_map_coordinates_from_pos(mouse_x, mouse_y, 0, map_x, map_y);
+
+        // Get viewport coordinates centring around the tile.
+        int32_t base_height = map::tile_element_height(*map_x, *map_y);
+        int16_t dest_x, dest_y;
+        viewport* v = this->viewports[0];
+        centre_2d_coordinates(*map_x, *map_y, base_height, &dest_x, &dest_y, v);
+
+        // Rebase mouse position onto centre of window, and compensate for zoom level.
+        int16_t rebased_x = ((this->width >> 1) - mouse_x) * (1 << v->zoom),
+                rebased_y = ((this->height >> 1) - mouse_y) * (1 << v->zoom);
+
+        // Compute cursor offset relative to tile.
+        viewport_config* vc = &this->viewport_configurations[0];
+        *offset_x = (vc->saved_view_x - (dest_x + rebased_x)) * (1 << v->zoom);
+        *offset_y = (vc->saved_view_y - (dest_y + rebased_y)) * (1 << v->zoom);
+    }
+
+    void window::viewport_centre_tile_around_cursor(int16_t map_x, int16_t map_y, int16_t offset_x, int16_t offset_y)
+    {
+        // Get viewport coordinates centring around the tile.
+        int16_t dest_x, dest_y;
+        int32_t base_height = map::tile_element_height(map_x, map_y);
+        viewport* v = this->viewports[0];
+        centre_2d_coordinates(map_x, map_y, base_height, &dest_x, &dest_y, v);
+
+        // Get mouse position to offset against.
+        int32_t mouse_x, mouse_y;
+        ui::get_cursor_pos(mouse_x, mouse_y);
```

Comment:
Not a fan of having these methods inside the functions. Would rather have them as a parameter for `viewport_zoom_in`

Reply:
Hmm, I get you'd like to leave the mouse functions out of these functions… But if I understand you correctly, you'd prefer to pass the mouse coords from `zoom_in` to `zoom_set` to these `centre` functions?